TUTORIAL.md

```diff
@@ -0,0 +1,392 @@
+    s.department_id = row[2]
+  end
+end
+```
+
+Now we can create a Courses table, insert a course and drop a course. Being able to find a course by it's name would also be useful, which is our next error. 
+
+We need to build a method called `find_by_name`. Again we can use `lib/student.rb` file as an example, replacing the necessary variables. Here we are selecting all from our courses table where the name is equal to the param we are passing in, then limiting it to one.
+
+```ruby
+ def self.find_by_name(name)
+    sql = <<-SQL
+      SELECT *
+      FROM courses
+      WHERE name = ?
+      LIMIT 1
+    SQL
+    result = DB[:conn].execute(sql,name)
+    result.map do |row| 
+      self.new_from_db(row)
+    end.first
+  end
+```
+Let's run `rspec` for our next error, `undefined method find_all_by_department_id`. Here we want to be able to find all courses by their `department_id`. The `SELECT` statement for this reads pretty clearly, `SELECT *` (`*` stands for all) from the courses table, where the `deparment_id` of the course equals the `id` we are passing in to the method, very simlilar to finding by name. Compare this code with your Student class to see the differences.
+
+```ruby
+def self.find_all_by_department_id(id)
+  sql = "SELECT * FROM courses WHERE department_id = ?"
+  result = DB[:conn].execute(sql,id)
+  result.collect do |row| 
+    self.new_from_db(row)
+  end
+end
+```
+Now we can create a Courses table, insert a course into the table, delete the table, find a course by name or department_id. What if we want to change the title of a course? We need a way to update the table.
+
+Run `rspec` and you will see `undefined method update`. Let's take a look at our `student.rb` file and see if there is anything to guide us. Looks like there is an `update` method built already, let's deconstruct it.
+
+The `update` method will take the values you are changing, find the object by it's id and update the instance. You can see our helper method `attribute_values` being passed in to the `DB[:conn].execute(sql, attribute_values, id)` part of the method.
+
+```ruby
+def update
+  sql = <<-SQL
+    UPDATE courses
+    SET name = ?,department_id = ?
+    WHERE id = ?
+  SQL
+  DB[:conn].execute(sql, attribute_values, id)
+end
+```
+The last piece of this puzzle is the ability to save our changes, which we will write two methods for that work together. 
+
+The first is `persisted?`. This is a helper method that checks whether or not there is already a row in the database for a given object. If it is persisted it returns true, if not it returns false. 
+
+In ruby, it is convention that methods ending in a `?` will return either true or false. 
+
+Skip down to the solution and you will notice the double exclamation points in the body of our `persisted?` method. These are called bangs. One bang will negate a boolean. So if something is true, a bang in front will make it false. If it is false, a bang in front will make it true. 
```

Comment:
I'm not super opinionated on this, but I hate calling excalamation points "bang". Too sexual